Stage 3 Authorization with React

skovakina opened the pull request:
# Implement Authorization Features in News Explorer (Stage 3)

## Description

This pull request addresses the implementation of Stage 3: Authorization with React. The focus is on integrating the React front-end with backend services to handle user authentication and secure access to specific routes.

## Major Changes

1. **Asynchronous API Requests**: Implemented asynchronous GET and POST requests to the backend, enabling dynamic user authentication and data retrieval in alignment with React's best practices.

2. **Route Protection**: Established protected routes within the application, ensuring that only authenticated users can access certain pages and data, enhancing security and user experience.

3. **Context API for Global State Management**: Utilized React's Context API to create and manage a global state variable `currentUser`, which stores user data across the application.

4. **LocalStorage for Token Management**: Implemented functionality to store and retrieve authentication tokens from localStorage, maintaining user sessions across page reloads.

5. **User Registration and Login Components**: Developed and integrated user registration and login components with front-end validation, providing clear feedback and instructions to users.

## Deployment Links

- Frontend: [News Explorer Frontend](https://sko-news.mooo.com/)
- Backend: [News Explorer API](https://api.sko-news.mooo.com/)

skovakina committed:
feat: sign in, sign up, logout

skovakina committed:
feat: curent user context

skovakina committed:
feat: saved articles

skovakina committed:
refactor: unused variables

skovakina committed:
feat: server url

skovakina committed:
fix: news api

skovakina committed:
Update README.md

gennady-bars commented on src/utils/auth.js:
```diff
@@ -0,0 +1,35 @@
+      email: data.email,
+      password: data.password,
+    }),
+  }).then((res) => checkResponse(res));
```
**Could be improved**

You can make a special method (function ) for fetching and checking responses **not to duplicate** it in every request:

```js

  function request(url, options) {
    return fetch(url, options).then(checkResponse)
  }

```

And now you can safely replace `fetch` with `request` and remove repeated checking. And you don’t need to change anything else.

gennady-bars commented on src/utils/serverApi.js:
```diff
@@ -1,29 +1,41 @@
 import checkResponse from './utils';
 
```
**Could be improved**

If it’s interesting for you here is a way to make a special wrapper for any `modal` with cross button, `Escape` and `overlay` listeners inside. You should create component `Modal`, store it in `Modal.js`. 


```jsx

import { useEffect } from "react";

export const Modal = ({ name, onClose, children }) => {
  // here is `useEffect` for the `Escape` listener
  useEffect(() => {
    // we should define the handler inside `useEffect`, so that it wouldn’t lose the reference to be able to remove it
    const handleEscape = (e) => {
      if (e.key === "Escape") {
        onClose();
      }
    };

    document.addEventListener("keydown", handleEscape);
    // don’t forget to remove the listener in the `clean-up` function
    return () => document.removeEventListener("keydown", handleEscape);
  }, [onClose]);

  // here is the overlay handler
  const handleOverlay = (e) => {
    if (e.target === e.currentTarget) {
      onClose();
    }
  };

  // then we add the main wrapper with class `modal`
  return (
    <div className={`modal modal_type_${name}`} onClick={handleOverlay}>
      {/* the container for the contents */}
      <div className="modal__container">
        {/* here will be anything you add as `children`*/}
        {children}
        {/* add the close button */}
        <button className="modal__close" type="button" onClick={onClose} />
      </div>
    </div>
  );
};



```


And now you can use `Modal` for any popup in the project:  `ItemModal`  and `ModalWithForm` 

```jsx


function ModalWithForm({ name, onClose, ...props}) {
  return (
    <Modal name={name} onClose={onClose}>
        <h2 className='popup__title'>{props.title}</h2>


```

You can make also `Input` for inputs and `Form` for any form in the project.

gennady-bars reviewed (commented):
Dear Sveta , (**Please, expand the general comment ↓**)

Good job but there are some remarks:



* I can save one and the same article a lot of times https://disk.yandex.ru/i/YlO4nHk1I78KGQ 
* The criteria say: When clicking on the active "Save" icon in the card block, a request to delete the card is made. After a successful request, the card is removed from the "Saved articles" page
* The criteria say: When an unauthorized user clicks on the icon to save an article, the registration popup window opens
* `UseFormValidation.js` is not a component. Please, move it into folder `hooks` outside `components`

Please, correct everything and your project will be accepted.

Good luck with the refactoring

## Number of points:  73

# Summary:
### Performance Criteria

- [] <!--22.04--> The project functionality is fully implemented according to the current stage's requirements:
  - General
    - All project links and buttons are functioning.
    - Both header states function correctly. If the user is not logged in, the header should have the "Sign in" button; and if the user is logged in, there should be no "Sign in" button. The "Saved articles" link, along with a "Log out" button, should appear in its place.
    - If the user closes the tab and then returns to the site, data is taken from local storage upon mounting the `App` component.
  - "Sign up" and "Sign in" pages
    - When clicking on the "Sign up" button in the "Sign up" popup window, a request is sent to the /signup route, provided that all input fields have been filled in correctly. If the request is successful, a popup should appear, which informs the user that they are registered and offers to log them in.
    - When clicking on the "Sign in" button, provided that all input fields have been filled in correctly, a request is sent to the `/signin` route. If the request is successful, the popup is closed.
    - Until there is valid data in all the input fields on the "Sign up" and "Sign in" pages, the form submit buttons remain disabled.
    - All forms are validated on the client side. The user cannot send a request with invalid data.
  - Search for cards
    - The preloader is visible and spinning during request execution.
    - Once the search form has been submitted successfully, a block with results appears. If nothing was found, the message "Nothing found" appears.
    - 3 cards are displayed in the results block. Clicking on the "Show more" button will render the next 3 cards.
    - **When an unauthorized user clicks on the icon to save an article, the registration popup window opens.**
    - After the user has been authorized, the "Save" icon becomes active in the card block.
    - When clicking the "Save" icon in the card block, a request is made to the /articles route of the API.
    - **When clicking on the active "Save" icon in the card block, a request to delete the card is made. After a successful request, the card is removed from the "Saved articles" page.**
  - "Saved Articles" page
    - The "Saved articles" page displays the following: username, number of articles saved, and the keywords by which articles were found.
    - Card blocks in the "Saved articles" page contain the keyword by which the card was found and the trash can icon for deleting the article.
    - When clicking on the trash icon of the card in the "Saved articles" page, a request to delete the card is made. After a successful request, the card is removed from the "Saved articles" page.
- Registration and authorization:
  - [x] <!--2.52--> At least one route is protected using the `ProtectedRoute` HOC component.
  - [] <!--2.52--> When trying to access the protected route, unauthorized users are redirected to / with an open authorization popup window.
  - [x] <!--2.52--> After successful authorization, the popup window is closed.
  - [x] <!--2.52--> The / route is not protected.
  - [x] <!--2.52--> If the user was logged in and closed the tab, they can return directly to any page of the application by the URL, except for the login and registration pages.
  - [x] <!--2.52--> After a successful onSignOut() handler call, the user is redirected to /.
  - [x] <!--2.52--> The `useHistory()` hook is used correctly.
  - [x] <!--2.52--> The components `<Switch />`, `<Route />`, and `<Redirect />` are used correctly.
- [x] The interaction with the JWT token works correctly:
  - [x] <!--2.52--> The JWT token is stored in localStorage.
  - [x] <!--2.52--> The JWT token is validated by a request to the server, not just local storage.
  - [x] <!--2.52--> When you log out of the account, JWT is deleted.
- [x] <!--2.52--> A global state variable has been created that stores user data.
- Components:
  - [x] <!--2.52--> Hooks are not used inside conditional statements or loops.
  - [x] <!--2.52--> Hooks are called in a component's main function.
  - [x] <!--2.52--> For class components, effects are described inside the component lifecycle methods.
  - [x] <!--2.52--> For list items, a unique id is used instead of an array index.
  - [x] <!--2.52--> Components that use profile data are subscribed to the context
  - [x] <!--2.52--> The context is embedded in the `App` component via `CurrentUserContext.Provider`.
  - [x] <!--2.52--> A state variable has been created in the root App component that stores user data. This variable is used as a value of the context provider.
- [x] <!--2.52--> Asynchronous API requests:
  - Requests can be made through the Fetch API or by using XMLHttpRequest. Third-party libraries (such as axios or jQuery) are not used.
  - API requests are contained in a separate file.
  - The chain for processing promises ends with a `catch()` block.
  - The first `then()` handler returns `res.json`.
- [x] <!--2.52--> Naming:
  - camelCase is used for function and variable names.
  - Only nouns are used as variable names.
  - Variable names clearly describe what is stored in them. If the project has several variables with similar data, then those variables have unique but descriptive names.
  - Descriptive names are used for functions, which reflect what they do.
  - Function names start with a verb.
  - JS classes and functional components are named using nouns and start with a capital letter.
  - Names must not include inappropriate or unclear abbreviations.
  - Сustom hook names start with `use`.
- [x] <!--2.52--> No third-party JavaScript libraries are used unless absolutely necessary. If third-party libraries are connected, then they are used appropriately.

### Best Practices

- [x] <!--2.5--> The initial state of state variables contain the correct data type.
- [x] <!--2.5--> The API request for information about the user and for the array of cards is made once during mounting.
- [x] <!--2.5--> API requests are described inside the App component.
- [x] <!--2.5--> There is no memory leak when hanging handlers. All handlers added with `addEventListener` are removed when the component is unmounted.
- [x] <!--2.5--> API error handling:
  - [] <!--2.5--> The user receives a message in the case of an error.
- [x] <!--2.5--> Non-variable values (hard-coded constants) are named in all capital letters and stored in a separate configuration file.

### Recommendations

- [x] <!--1.66--> Components from the react-router library are used for internal links in the application. 
- [x] <!--1.66--> Semantically correct blocks are used for components. No `<div />` or other unnecessary HTML tags are used for components that consist of single-level blocks.
- [x] <!--1.68--> The code is clean and easy to understand:
  - The code is readable and clearly structured. Some parts of the code are explained with comments if needed.
  - There is no extra code: for example, when a variable is declared but not used, or there is some kind of redundant logic.
  - The code is formatted in the same way, and the indentation hierarchy is respected.

## Number of points:  73

## Additional comments:
-

skovakina committed:
fix: news card can be saved only once

skovakina committed:
fix: Update article save/delete functionality

- Articles can only be saved once
- Clicking on an active "Save" icon now deletes the article from "Saved articles".
- Unauthorized user clicks on save icon now triggers registration popup.
- Moved UseFormValidation.js to hooks folder.
- Updated article card schema for consistent rendering across the app.

skovakina committed:
Merge branch 'stage-3-auth' of https://github.com/skovakina/se_news-explore into stage-3-auth

skovakina commented:
Hello! Thank you for your review!
Implemented enhancements for user interactions with articles and optimized code organization:
- Articles can now only be saved once to prevent duplicates.
- Toggling the "Save" icon on articles now correctly deletes them from saved lists.
- Unregistered users attempting to save articles are prompted to sign up.
- UseFormValidation.js has been relocated to the hooks folder for improved code organization.
- Updated article card schema for uniform rendering across the platform.

gennady-bars reviewed (approved):
You’ve done a great job.

Your project has been accepted. 

Good luck

## Number of points: 95

# Summary:
### Performance Criteria

- [x] <!--22.04--> The project functionality is fully implemented according to the current stage's requirements:
  - General
    - All project links and buttons are functioning.
    - Both header states function correctly. If the user is not logged in, the header should have the "Sign in" button; and if the user is logged in, there should be no "Sign in" button. The "Saved articles" link, along with a "Log out" button, should appear in its place.
    - If the user closes the tab and then returns to the site, data is taken from local storage upon mounting the `App` component.
  - "Sign up" and "Sign in" pages
    - When clicking on the "Sign up" button in the "Sign up" popup window, a request is sent to the /signup route, provided that all input fields have been filled in correctly. If the request is successful, a popup should appear, which informs the user that they are registered and offers to log them in.
    - When clicking on the "Sign in" button, provided that all input fields have been filled in correctly, a request is sent to the `/signin` route. If the request is successful, the popup is closed.
    - Until there is valid data in all the input fields on the "Sign up" and "Sign in" pages, the form submit buttons remain disabled.
    - All forms are validated on the client side. The user cannot send a request with invalid data.
  - Search for cards
    - The preloader is visible and spinning during request execution.
    - Once the search form has been submitted successfully, a block with results appears. If nothing was found, the message "Nothing found" appears.
    - 3 cards are displayed in the results block. Clicking on the "Show more" button will render the next 3 cards.
    - When an unauthorized user clicks on the icon to save an article, the registration popup window opens.
    - After the user has been authorized, the "Save" icon becomes active in the card block.
    - When clicking the "Save" icon in the card block, a request is made to the /articles route of the API.
    - When clicking on the active "Save" icon in the card block, a request to delete the card is made. After a successful request, the card is removed from the "Saved articles" page.
  - "Saved Articles" page
    - The "Saved articles" page displays the following: username, number of articles saved, and the keywords by which articles were found.
    - Card blocks in the "Saved articles" page contain the keyword by which the card was found and the trash can icon for deleting the article.
    - When clicking on the trash icon of the card in the "Saved articles" page, a request to delete the card is made. After a successful request, the card is removed from the "Saved articles" page.
- Registration and authorization:
  - [x] <!--2.52--> At least one route is protected using the `ProtectedRoute` HOC component.
  - [] <!--2.52--> When trying to access the protected route, unauthorized users are redirected to / with an open authorization popup window.
  - [x] <!--2.52--> After successful authorization, the popup window is closed.
  - [x] <!--2.52--> The / route is not protected.
  - [x] <!--2.52--> If the user was logged in and closed the tab, they can return directly to any page of the application by the URL, except for the login and registration pages.
  - [x] <!--2.52--> After a successful onSignOut() handler call, the user is redirected to /.
  - [x] <!--2.52--> The `useHistory()` hook is used correctly.
  - [x] <!--2.52--> The components `<Switch />`, `<Route />`, and `<Redirect />` are used correctly.
- [x] The interaction with the JWT token works correctly:
  - [x] <!--2.52--> The JWT token is stored in localStorage.
  - [x] <!--2.52--> The JWT token is validated by a request to the server, not just local storage.
  - [x] <!--2.52--> When you log out of the account, JWT is deleted.
- [x] <!--2.52--> A global state variable has been created that stores user data.
- Components:
  - [x] <!--2.52--> Hooks are not used inside conditional statements or loops.
  - [x] <!--2.52--> Hooks are called in a component's main function.
  - [x] <!--2.52--> For class components, effects are described inside the component lifecycle methods.
  - [x] <!--2.52--> For list items, a unique id is used instead of an array index.
  - [x] <!--2.52--> Components that use profile data are subscribed to the context
  - [x] <!--2.52--> The context is embedded in the `App` component via `CurrentUserContext.Provider`.
  - [x] <!--2.52--> A state variable has been created in the root App component that stores user data. This variable is used as a value of the context provider.
- [x] <!--2.52--> Asynchronous API requests:
  - Requests can be made through the Fetch API or by using XMLHttpRequest. Third-party libraries (such as axios or jQuery) are not used.
  - API requests are contained in a separate file.
  - The chain for processing promises ends with a `catch()` block.
  - The first `then()` handler returns `res.json`.
- [x] <!--2.52--> Naming:
  - camelCase is used for function and variable names.
  - Only nouns are used as variable names.
  - Variable names clearly describe what is stored in them. If the project has several variables with similar data, then those variables have unique but descriptive names.
  - Descriptive names are used for functions, which reflect what they do.
  - Function names start with a verb.
  - JS classes and functional components are named using nouns and start with a capital letter.
  - Names must not include inappropriate or unclear abbreviations.
  - Сustom hook names start with `use`.
- [x] <!--2.52--> No third-party JavaScript libraries are used unless absolutely necessary. If third-party libraries are connected, then they are used appropriately.

### Best Practices

- [x] <!--2.5--> The initial state of state variables contain the correct data type.
- [x] <!--2.5--> The API request for information about the user and for the array of cards is made once during mounting.
- [x] <!--2.5--> API requests are described inside the App component.
- [x] <!--2.5--> There is no memory leak when hanging handlers. All handlers added with `addEventListener` are removed when the component is unmounted.
- [x] <!--2.5--> API error handling:
  - [] <!--2.5--> The user receives a message in the case of an error.
- [x] <!--2.5--> Non-variable values (hard-coded constants) are named in all capital letters and stored in a separate configuration file.

### Recommendations

- [x] <!--1.66--> Components from the react-router library are used for internal links in the application. 
- [x] <!--1.66--> Semantically correct blocks are used for components. No `<div />` or other unnecessary HTML tags are used for components that consist of single-level blocks.
- [x] <!--1.68--> The code is clean and easy to understand:
  - The code is readable and clearly structured. Some parts of the code are explained with comments if needed.
  - There is no extra code: for example, when a variable is declared but not used, or there is some kind of redundant logic.
  - The code is formatted in the same way, and the indentation hierarchy is respected.

## Number of points: 95

## Additional comments:
-

skovakina committed:
Merge branch 'main' into stage-3-auth
